Resolves #51 "[Groups] Remove a user from a Group"

mortimr opened the pull request:
Resolves #51

benard-g commented on sources/api/controllers/GroupController.ts:
```diff
@@ -106,3 +106,66 @@ export async function deleteGroup(req: Request, res: Response) {
+        let user_id: number = (await getUser(userRepository, user)).id;
+
+        // Check if user is a member of the group
+        if (group.members.map((e: User) => e.id).indexOf(user_id) === -1)
```
Do not do the `map()` every time, store the id list in a `Set`

benard-g commented on sources/api/controllers/GroupController.ts:
```diff
@@ -106,3 +106,66 @@ export async function deleteGroup(req: Request, res: Response) {
+
+    // Has the owner been removed ? Find a new one
+    if (group.members.findIndex((e: User) => e.id === group.owner.id) === -1) {
+        group.owner = group.members[Math.floor(Math.random() * group.members.length)];
```
Just take the next in the queue 😕 (first of the list).

benard-g commented on sources/api/controllers/GroupController.ts:
```diff
@@ -106,3 +106,66 @@ export async function deleteGroup(req: Request, res: Response) {
+        group.owner = group.members[Math.floor(Math.random() * group.members.length)];
+    }
+
+    await conn.manager.save(group);
```
Use the repository since you have it

benard-g commented on sources/api/controllers/GroupController.ts:
```diff
@@ -106,3 +106,66 @@ export async function deleteGroup(req: Request, res: Response) {
+        if (group.members.map((e: User) => e.id).indexOf(user_id) === -1)
+            throw new ApiException(404, `User ${user} is not a member of the group.`);
+
+        // Check if calling user has proper rights to remove other users
```
```suggestion
        // Check if calling user is removing itself
        // or has proper rights to remove other users
```

I think it's clearer this way (at least this is what I think you are trying to do here)

benard-g commented on sources/api/routes/groups/index.ts:
```diff
@@ -15,4 +15,6 @@ router.get('/:group_id', AsyncHandler(loggedOnly), AsyncHandler(GroupController.
 
 router.delete('/:group_id', AsyncHandler(loggedOnly), AsyncHandler(GroupController.deleteGroup));
 
+router.delete('/kick/:group_id', AsyncHandler(loggedOnly), AsyncHandler(GroupController.kickFromGroup));
```
'/:group_id/kick'

benard-g commented:
@mortimr Close this PR

mortimr commented:
@benard-g I will force-push the new branch with this name to keep this PR

mortimr committed:
Remove User Implementation

- removeUser method added to Calendar entity
- kickUserFromGroup method added to GroupController
- findByCalendarId method added to Group entity

API #51

mortimr committed:
Updated test suite

API #51

aliphe commented on sources/model/entity/Group.ts:
```diff
@@ -57,4 +74,5 @@ export class Group {
             .where({id: groupId})
             .execute();
     }
+
```
Beware of the Benard

benard-g commented on sources/model/entity/Calendar.ts:
```diff
@@ -187,12 +188,51 @@ export class Calendar {
+        const cto: CalendarsToOwners = await CalendarsToOwners.findCalendarRelation(conn, calendarId, user.id);
+
+        // Remove relation
+        await calendarsToOwnerRepository.remove(cto);
```
Use `QueryBuilder` to directly remove the relationship

```ts
// Something like
queryBuilder.delete().where("calendar_id").andWhere("owner_id").execute();
```

benard-g commented on sources/model/entity/Calendar.ts:
```diff
@@ -187,12 +188,51 @@ export class Calendar {
+        await calendarsToOwnerRepository.remove(cto);
+
+        const members: CalendarsToOwners[] = await Calendar.getCalendarOwners(conn, calendarId);
+        if (members.length == 0) {
```
Get the Calendar directly, you can use the field `nb_owners`

benard-g commented on sources/model/entity/Calendar.ts:
```diff
@@ -187,12 +188,51 @@ export class Calendar {
+            const group: Group = await Group.findByCalendarId(conn, calendarId);
+
+            // Delete group linked to empty calendar
+            await Group.deleteById(conn, group.id);
```
Be carreful, all Calendars aren't linked to a Group

benard-g commented on sources/model/entity/Calendar.ts:
```diff
@@ -187,12 +188,54 @@ export class Calendar {
+                .createQueryBuilder()
+                .delete()
+                .where("nb_owners = 0")
+                .andWhere("id = :calendar_id", {calendar_id: calendarId})
```
No need to check `nb_owners` you have the `id`

benard-g commented on sources/api/controllers/GroupController.ts:
```diff
@@ -260,22 +260,50 @@ export async function editUserStatus(req: Request, res: Response) {
+    // Remove user from calendar
+    await Calendar.removeUser(conn, group.calendar.id, rm_user);
+
+    // TODO: Put this in the View once we manage roles
```
👍
